pkg/nsx/services/subnetport/subnetport.go

```diff
@@ -0,0 +1,222 @@
+/* Copyright © 2023 VMware, Inc. All Rights Reserved.
+   SPDX-License-Identifier: Apache-2.0 */
+
+package subnetport
+
+import (
+	"context"
+	"errors"
+	"strings"
+
+	"sync"
+
+	"github.com/vmware/vsphere-automation-sdk-go/services/nsxt/model"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/tools/cache"
+
+	"github.com/vmware-tanzu/nsx-operator/pkg/logger"
+	"github.com/vmware-tanzu/nsx-operator/pkg/nsx/services/common"
+	"github.com/vmware-tanzu/nsx-operator/pkg/nsx/util"
+
+	"github.com/vmware-tanzu/nsx-operator/pkg/apis/v1alpha1"
+)
+
+var (
+	log                    = logger.Log
+	ResourceTypeSubnetPort = common.ResourceTypeSubnetPort
+	MarkedForDelete        = true
+)
+
+type SubnetPortService struct {
+	common.Service
+	SubnetPortStore *SubnetPortStore
+}
+
+// InitializeSubnetPort sync NSX resources
+func InitializeSubnetPort(service common.Service) (*SubnetPortService, error) {
+	wg := sync.WaitGroup{}
+	wgDone := make(chan bool)
+	fatalErrors := make(chan error)
+
+	wg.Add(1)
+
+	subnetPortService := &SubnetPortService{Service: service}
+
+	subnetPortService.SubnetPortStore = &SubnetPortStore{ResourceStore: common.ResourceStore{
+		Indexer:     cache.NewIndexer(keyFunc, cache.Indexers{common.TagScopeSubnetPortCRUID: subnetPortIndexFunc}),
+		BindingType: model.SegmentPortBindingType(),
+	}}
+
+	go subnetPortService.InitializeResourceStore(&wg, fatalErrors, ResourceTypeSubnetPort, nil, subnetPortService.SubnetPortStore)
+
+	go func() {
+		wg.Wait()
+		close(wgDone)
+	}()
+
+	select {
+	case <-wgDone:
+		break
+	case err := <-fatalErrors:
+		close(fatalErrors)
+		return subnetPortService, err
+	}
+
+	return subnetPortService, nil
+}
+
+func (service *SubnetPortService) CreateOrUpdateSubnetPort(obj *v1alpha1.SubnetPort) error {
+
+	log.Info("Creating or updating subnetport", "subnetport", obj)
+
+	// TODO: For now, only the parent subnet is supported.
+	nsxSubnetPath, err := service.GetPathFromSubnet(obj)
+	if err != nil {
+		log.Error(err, "failed to get NSX resource path from subnet", "subnetport", obj)
+		return err
+	}
+
+	nsxSubnetPort, err := service.buildSubnetPort(obj)
+	if err != nil {
+		log.Error(err, "failed to build NSX subnet port")
+		return err
+	}
+
+	subnetInfo, err := common.ParseVPCResourcePath(nsxSubnetPath)
+	if err != nil {
+		log.Error(err, "failed to parse NSX subnet path", "path", nsxSubnetPath)
+	}
+	existingSubnetPort := service.SubnetPortStore.GetByKey(*nsxSubnetPort.Id)
+	isChanged := common.CompareResource(SubnetPortToComparable(existingSubnetPort), SubnetPortToComparable(nsxSubnetPort))
```

Comment:
if there is no existingSubnetPort, can we still compare the resource?

Reply:
yes, if it's not in the store, the `existingSubnetPort ` will be `nil` and `isChanged` will be `true`